### PR TITLE
llama : reorganize source code + improve CMake 

### PR DESCRIPTION
**Pull request by ggerganov**
ref #7573, #6913 

Adopt a new structure of the source code that makes `ggml` source code and build scripts easier to reuse.

Note that the build options that are relevant to the `ggml` library are now prefixed with `GGML_`. For example, `LLAMA_CUDA` and `WHISPER_METAL` are now `GGML_CUDA` and `GGML_METAL`. However, `WHISPER_COREML` and `WHISPER_OPENVINO` are still named the same because the CoreML and OpenVINO functionality in `whisper.cpp` is not part of the `ggml` library.

The `Makefiles` in `llama.cpp` and `whisper.cpp` have been updated to be more similar with each other.

Header files (such as `ggml.h`, `llama.h` and `whisper.h`) are now placed in `include` subfolders, while the source files are placed in `src`.

PRs in other projects that will be updated and merged together with this one:

- https://github.com/ggerganov/ggml/pull/865
- https://github.com/ggerganov/whisper.cpp/pull/2256

TODOs:

- [x] update `ggml` repo
- [x] update `whisper.cpp` repo
- [x] update Makefiles
  - [x] `llama.cpp`
  - [x] `whisper.cpp`
- [x] update sync scripts
  - [x] `llama.cpp`
  - [x] `ggml`
  - [x] `whisper.cpp`
  - [x] add `ggml/cmake`
- [x] fix examples
- [x] fix CI
  - [x] HIPBLAS
  - [x] kompute
  - [x] nix (maybe will work after merge to `master`?)
  - [x] windows
- [x] update README with new `GGML_` options
  - [x] `llama.cpp`
  - [x] `whisper.cpp`

Changes:

- deprecate `LLAMA_XXX` in favor of `GGML_XXX`

Resolved PRs:

- #8003
- #8017
- #8075

TODO in follow-up PRs:

- [ ] move relevant tests from `tests` to `ggml/tests`
- [ ] avoid using `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS` and build and link shared libraries properly
- [ ] link shared libs to examples by default
- [ ] simplify sync scripts to work on the `ggml` folder level

---

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [ ] Low
  - [ ] Medium
  - [ ] High

### CODE REVIEW

**Comment by ggerganov** [comment]
The `llama.cpp` reorganization should be now mostly ready and can be tested. The `ggml` code, tests and cmake scripts will be shared across all 3 repos (as if it is a git submodule).

Some CI workflows are still failing - any help with resolving these is appreciated. I'll now focus on updating `whisper.cpp` to reuse `ggml` in a similar way and then look to merge this sometime next week. Extra attention to the new build options (e.g. `LLAMA_CUDA` is now `GGML_CUDA`)

**Comment by slaren** [comment]
The windows build is failing because some targets do not link to the libraries they need. On linux that works because in the end all the symbols all put together indiscriminately, but not on windows. 

This solves some of the problems, but some tests depend on internal functions such as `unicode_cpt_to_utf8` of llama.cpp that are not exported. But I am not sure if this a real solution, because with static builds I think this is going to cause multiple versions of ggml and llama to be linked in some binaries.


```diff
diff --git a/common/CMakeLists.txt b/common/CMakeLists.txt
index c6fccc02..299d53c3 100644
--- a/common/CMakeLists.txt
+++ b/common/CMakeLists.txt
@@ -85,4 +85,4 @@ endif ()

 target_include_directories(${TARGET} PUBLIC .)
 target_compile_features   (${TARGET} PUBLIC cxx_std_11)
-target_link_libraries     (${TARGET} PRIVATE ${LLAMA_COMMON_EXTRA_LIBS} PUBLIC llama Threads::Threads)
+target_link_libraries     (${TARGET} PRIVATE ${LLAMA_COMMON_EXTRA_LIBS} PUBLIC ggml llama Threads::Threads)
diff --git a/examples/benchmark/CMakeLists.txt b/examples/benchmark/CMakeLists.txt
index 34a58cc0..358fd130 100644
--- a/examples/benchmark/CMakeLists.txt
+++ b/examples/benchmark/CMakeLists.txt
@@ -1,6 +1,6 @@
 set(TARGET llama-bench-matmult)
 add_executable(${TARGET} benchmark-matmult.cpp)
 install(TARGETS ${TARGET} RUNTIME)
-target_link_libraries(${TARGET} PRIVATE llama build_info ${CMAKE_THREAD_LIBS_INIT})
+target_link_libraries(${TARGET} PRIVATE ggml llama build_info ${CMAKE_THREAD_LIBS_INIT})
 target_include_directories(${TARGET} PRIVATE ../../common)
 target_compile_features(${TARGET} PRIVATE cxx_std_11)
diff --git a/examples/gbnf-validator/CMakeLists.txt b/examples/gbnf-validator/CMakeLists.txt
index 4edd6ec7..29530835 100644
--- a/examples/gbnf-validator/CMakeLists.txt
+++ b/examples/gbnf-validator/CMakeLists.txt
@@ -1,5 +1,5 @@
 set(TARGET llama-gbnf-validator)
 add_executable(${TARGET} gbnf-validator.cpp)
 install(TARGETS ${TARGET} RUNTIME)
-target_link_libraries(${TARGET} PRIVATE common llama ${CMAKE_THREAD_LIBS_INIT})
+target_link_libraries(${TARGET} PRIVATE common ggml llama ${CMAKE_THREAD_LIBS_INIT})
 target_compile_features(${TARGET} PRIVATE cxx_std_11)
diff --git a/examples/llama-bench/CMakeLists.txt b/examples/llama-bench/CMakeLists.txt
index 5bdbea4e..eb1918c2 100644
--- a/examples/llama-bench/CMakeLists.txt
+++ b/examples/llama-bench/CMakeLists.txt
@@ -1,5 +1,5 @@
 set(TARGET llama-bench)
 add_executable(${TARGET} llama-bench.cpp)
 install(TARGETS ${TARGET} RUNTIME)
-target_link_libraries(${TARGET} PRIVATE common llama ${CMAKE_THREAD_LIBS_INIT})
+target_link_libraries(${TARGET} PRIVATE ggml common llama ${CMAKE_THREAD_LIBS_INIT})
 target_compile_features(${TARGET} PRIVATE cxx_std_11)
diff --git a/examples/quantize-stats/CMakeLists.txt b/examples/quantize-stats/CMakeLists.txt
index bb986a71..73e7f943 100644
--- a/examples/quantize-stats/CMakeLists.txt
+++ b/examples/quantize-stats/CMakeLists.txt
@@ -1,6 +1,6 @@
 set(TARGET llama-quantize-stats)
 add_executable(${TARGET} quantize-stats.cpp)
 install(TARGETS ${TARGET} RUNTIME)
-target_link_libraries(${TARGET} PRIVATE llama build_info ${CMAKE_THREAD_LIBS_INIT})
+target_link_libraries(${TARGET} PRIVATE ggml llama build_info ${CMAKE_THREAD_LIBS_INIT})
 target_include_directories(${TARGET} PRIVATE ../../common)
 target_compile_features(${TARGET} PRIVATE cxx_std_11)
diff --git a/include/llama.h b/include/llama.h
index 05d8b092..d833ba4a 100644
--- a/include/llama.h
+++ b/include/llama.h
@@ -1124,23 +1124,23 @@ struct llama_grammar_candidate {
     llama_partial_utf8   partial_utf8;
 };

-const std::vector<std::pair<std::string, struct ggml_tensor *>> & llama_internal_get_tensor_map(
+LLAMA_API const std::vector<std::pair<std::string, struct ggml_tensor *>> & llama_internal_get_tensor_map(
     struct llama_context * ctx
 );

-void llama_grammar_accept(
+LLAMA_API void llama_grammar_accept(
         const std::vector<std::vector<llama_grammar_element>>         & rules,
         const std::vector<std::vector<const llama_grammar_element *>> & stacks,
         const uint32_t                                                  chr,
         std::vector<std::vector<const llama_grammar_element *>>       & new_stacks);

-std::pair<std::vector<uint32_t>, llama_partial_utf8> decode_utf8(
+LLAMA_API std::pair<std::vector<uint32_t>, llama_partial_utf8> decode_utf8(
         const std::string & src,
         llama_partial_utf8   partial_start);

 // Randomly selects a token from the candidates based on their probabilities using given std::mt19937.
 // This is a temporary workaround in order to fix race conditions when sampling with multiple sequences.
-llama_token llama_sample_token_with_rng(struct llama_context * ctx, llama_token_data_array * candidates, std::mt19937 & rng);
+LLAMA_API llama_token llama_sample_token_with_rng(struct llama_context * ctx, llama_token_data_array * candidates, std::mt19937 & rng);

 #endif // LLAMA_API_INTERNAL

diff --git a/tests/CMakeLists.txt b/tests/CMakeLists.txt
index cfa70731..f70cdb2b 100644
--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -55,7 +55,7 @@ function(llama_target_and_test source)

     add_executable(${TEST_TARGET} ${source} get-model.cpp)
     install(TARGETS ${TEST_TARGET} RUNTIME)
-    target_link_libraries(${TEST_TARGET} PRIVATE common)
+    target_link_libraries(${TEST_TARGET} PRIVATE ggml llama common)
     add_test(
         NAME ${TEST_TARGET}
         WORKING_DIRECTORY ${LLAMA_TEST_WORKING_DIRECTORY}
```

**Comment by slaren** [comment]
It works with `-DBUILD_SHARED_LIBS=OFF`, so I am probably misunderstanding the error.

**Comment by slaren** [comment]
I noticed now that `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS` is set in ggml when building on windows with shared libs. I think that the problem actually is that `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS` is not set early enough, because configuring with `-DCMAKE_WINDOWS_EXPORT_ALL_SYMBOLS=ON` also works.

File: CMakeLists.txt
Comment (slaren) [inline CMakeLists.txt:72]: `GGML_BUILD_SHARED_LIBS` and `LLAMA_BUILD_SHARED_LIBS` do not exists, `BUILD_SHARED_LIBS` is used directly.

**Comment by mofosyne** [comment]
if you are reorganizing source codes, I would like to suggest moving some compiled tools from the example folder into a dedicated folder. It's not quite a script, but in my opinion its not examples as well... considering it's more for maintainers internal usage

**Comment by ggerganov** [comment]
The PRs are ready for review. The CI in `llama.cpp` is still failing for some of the workflows:

- I think HIPBLAS is failing because with `enable_language(HIP)` inside `ggml/src/CMakeLists.txt` instead of the root `CMakeLists.txt`. This is causing the undefined `CMAKE_HIP_` variables later on: https://github.com/ggerganov/llama.cpp/actions/runs/9657886762/job/26637996149?pr=8006. Not sure how to fix
- Hopefully the Nix workflows will work after merge because I think they are pulling stuff from `master`?
- Not sure why the `kompute` build is failing - @cebtenzzre could you assist with this? Should be a simple fix, but wasn't able to setup the `kompute` build on my machine
- The `whisper.cpp clang linux/arm/v7` workflow no longer works, though it looks like a `clang` problem: https://github.com/ggerganov/whisper.cpp/actions/runs/9657764109/job/26637633042?pr=2256#step:4:1990

There are quite some things to improve upon this, but probably it is better to merge first and resolve from `master`.

**Comment by slaren** [comment]
> * I think HIPBLAS is failing because with `enable_language(HIP)` inside `ggml/src/CMakeLists.txt` instead of the root `CMakeLists.txt`. This is causing the undefined `CMAKE_HIP_` variables later on: https://github.com/ggerganov/llama.cpp/actions/runs/9657886762/job/26637996149?pr=8006. Not sure how to fix

This may be related to this: https://cmake.org/cmake/help/latest/policy/CMP0165.html#policy:CMP0165

It may also affect CUDA. I get these warnings:
```
Make Warning (dev) in src/CMakeLists.txt:
  Policy CMP0104 is not set: CMAKE_CUDA_ARCHITECTURES now detected for NVCC,
  empty CUDA_ARCHITECTURES not allowed.  Run "cmake --help-policy CMP0104"
  for policy details.  Use the cmake_policy command to set the policy and
  suppress this warning.

  CUDA_ARCHITECTURES is empty for target "llama".
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) in examples/gguf/CMakeLists.txt:
  Policy CMP0104 is not set: CMAKE_CUDA_ARCHITECTURES now detected for NVCC,
  empty CUDA_ARCHITECTURES not allowed.  Run "cmake --help-policy CMP0104"
  for policy details.  Use the cmake_policy command to set the policy and
  suppress this warning.

  CUDA_ARCHITECTURES is empty for target "llama-gguf".
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) in examples/llava/CMakeLists.txt:
  Policy CMP0104 is not set: CMAKE_CUDA_ARCHITECTURES now detected for NVCC,
  empty CUDA_ARCHITECTURES not allowed.  Run "cmake --help-policy CMP0104"
  for policy details.  Use the cmake_policy command to set the policy and
  suppress this warning.

  CUDA_ARCHITECTURES is empty for target "llava_shared".
This warning is for project developers.  Use -Wno-dev to suppress it.
```

I am not sure what is supposed to be the solution in this case. I think that the source of the problem is using multiple `CMakeLists.txt` for the same target. Maybe this could be fixed by using `include` instead of `add_subdirectory`.

**Comment by slaren** [comment]
Probably the simplest solution would be to remove the `src/CMakeLists.txt` entirely and marge them into the base `CMakeList.txt`. I think that the expectation is that `add_subdirectory` is only used when a project has multiple targets.

**Comment by ggerganov** [comment]
> Probably the simplest solution would be to remove the src/CMakeLists.txt entirely and merge them into the base CMakeList.txt.

We still need to `add_subdirectory(ggml)` in the root `llama.cpp/CMakeLists.txt` - wouldn't it hit the same issue?

I've resolved these warnings in `whisper.cpp` by enabling CUDA in `whisper.cpp/src/CMakeLists.txt`: https://github.com/ggerganov/whisper.cpp/blob/f3c609bd2e69b40f7500af8e13c1db270f997403/src/CMakeLists.txt#L80-L101

We can do the same in `llama.cpp/src/CMakeLists.txt`, but for HIP it seems more complicated because it checks for some compiler compatibility

**Comment by slaren** [comment]
> We still need to `add_subdirectory(ggml)` in the root `llama.cpp/CMakeLists.txt` - wouldn't it hit the same issue?

My expectation was that ggml would be built either as a static or a shared library, and then linked to llama.cpp. However this does not seem to be happening because the `ggml` target is created as an "object" library, which seems to mean that it does not behave as a library at all, and it is more like adding the source files to the parent project. Changing this fixes the CUDA warnings for me.
```diff
index 4f146026..f7ecdc2a 100644
--- a/ggml/src/CMakeLists.txt
+++ b/ggml/src/CMakeLists.txt
@@ -1129,7 +1129,7 @@ endif()

 # ggml

-add_library(ggml OBJECT
+add_library(ggml
             ../include/ggml.h
             ../include/ggml-alloc.h
             ../include/ggml-backend.h
```

Is there any reason to use an `OBJECT` library?

**Comment by ggerganov** [comment]
Ah, it wasn't intended to be an object library - I just didn't notice. Nice catch

File: ggml/src/CMakeLists.txt
Comment (slaren) [inline ggml/src/CMakeLists.txt:1173]: This is probably not necessary here, it is already done in line 1164.

Comment (ggerganov) [reply ggml/src/CMakeLists.txt]: Force-pushed a fix and I also think the Kompute build should work now

**Comment by slaren** [comment]
This may also fix ggml being linked multiple times when using shared libraries, which should reduce the size of the pre-build binaries significantly. Currently, the windows CUDA binaries are almost 1 GB in total. Compare the file sizes of my build from this branch with the latest build from github:

image

File: Makefile
Comment (slaren) [inline Makefile:191]: The removal of this is causing the `make` build to be built without optimizations.

**Comment by slaren** [comment]
There are still some references to `LLAMA_CUDA` in the `Makefile`:
```diff
diff --git a/Makefile b/Makefile
index 88084cac..80e7678d 100644
--- a/Makefile
+++ b/Makefile
@@ -859,7 +859,7 @@ override NVCCFLAGS := $(MK_NVCCFLAGS) $(NVCCFLAGS)
 override LDFLAGS   := $(MK_LDFLAGS) $(LDFLAGS)

 # identify CUDA host compiler
-ifdef LLAMA_CUDA
+ifdef GGML_CUDA
 GF_CC := $(NVCC) $(NVCCFLAGS) 2>/dev/null .c -Xcompiler
 include scripts/get-flags.mk
 CUDA_CXXFLAGS := $(BASE_CXXFLAGS) $(GF_CXXFLAGS) -Wno-pedantic
@@ -884,7 +884,7 @@ $(info I NVCCFLAGS: $(NVCCFLAGS))
 $(info I LDFLAGS:   $(LDFLAGS))
 $(info I CC:        $(shell $(CC)   --version | head -n 1))
 $(info I CXX:       $(shell $(CXX)  --version | head -n 1))
-ifdef LLAMA_CUDA
+ifdef GGML_CUDA
 $(info I NVCC:      $(shell $(NVCC) --version | tail -n 1))
 CUDA_VERSION := $(shell $(NVCC) --version | grep -oP 'release (\K[0-9]+\.[0-9])')
 ifeq ($(shell awk -v "v=$(CUDA_VERSION)" 'BEGIN { print (v < 11.7) }'),1)
@@ -896,7 +896,7 @@ endif # CUDA_POWER_ARCH
 endif # CUDA_DOCKER_ARCH

 endif # eq ($(shell echo "$(CUDA_VERSION) < 11.7" | bc),1)
-endif # LLAMA_CUDA
+endif # GGML_CUDA
 $(info )

 ifdef DEPRECATE_WARNING
```

File: ggml/src/CMakeLists.txt
Comment (slaren) [inline ggml/src/CMakeLists.txt:314]: I am not sure if `GGML_CDEF_PRIVATE` is really necessary, or I don't understand its purpose. I think that using `add_compile_definitions` would achieve the same thing. From what I can tell, both `target_compile_definitions(PRIVATE)` and `add_compile_definitions` modify the same underlying property `COMPILE_DEFINITIONS`. For the public definitions it's different since `INTERFACE_COMPILE_DEFINITIONS` needs to be modified as well, so that in that case I think that `GGML_CDEF_PUBLIC` makes sense.

There are also a lot of usages of `add_compile_options` in this file, so if this is really necessary, shouldn't all of these be replaced as well?

Otherwise this should be updated to follow the same pattern:
```suggestion
            list(APPEND GGML_CDEF_PRIVATE GGML_CUDA_FORCE_CUBLAS)
```

Comment (ggerganov) [reply ggml/src/CMakeLists.txt:313]: Restored `add_compile_definitions` for the private flags

**Review by slaren — APPROVED** [review]

**Comment by OuadiElfarouki** [comment]
@slaren The `LLAMA_CUDA_FORCE_CUBLAS` cmake option got mistakenly removed but is still used. I believe it's intended to be mutually exclusive with `GGML_CUDA_FORCE_MMQ` so some changes might be needed (cmake, mmq.cu, ggml-cuda.cu ..)

**Comment by ggerganov** [comment]
There is now `GGML_CUDA_FORCE_CUBLAS`

**Pull request merged**
